### PR TITLE
P2609R3: Relaxing Ranges Just A Smidge

### LINKED ISSUE #3441: P2609R3 Relaxing Ranges Just A Smidge
WG21-P2609R3 Relaxing Ranges Just A Smidge

**Increased** feature-test macro (expected):
```cpp
#define __cpp_lib_ranges 202302L
```

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
Closes #3441.

Drive-by: reduced clang-format suppressed regions.

### CODE REVIEW

File: stl/inc/xutility [hunk 1]
```diff
@@ -701,6 +697,13 @@ struct projected {
     }
 };
 
+#if _HAS_CXX23
+template <class _It, class _Proj>
+struct _Indirect_value_impl<projected<_It, _Proj>> {
+    using type = invoke_result_t<_Proj&, _Indirect_value_t<_It>>;
+};
+#endif // _HAS_CXX23
+
```
Comment (JMazurkiewicz) [inline stl/inc/xutility:706]: To be clear: `_Indirect_value_t<T>` in C++20 mode is same as `iter_value_t<T>&`, and only in C++23 mode we add this extra specialization.

File: tests/std/tests/P2609R3_relaxing_ranges_just_a_smidge/test.compile.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <iterator>
```
Comment (CaseyCarter) [inline tests/std/tests/P2609R3_relaxing_ranges_just_a_smidge/test.compile.pass.cpp:5]: `<ranges>` must include `<iterator>` which must include `<concepts>`.
```suggestion
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/yvals_core.h [hunk 3]
```diff
@@ -352,6 +352,7 @@
 // P2505R5 Monadic Functions For expected
 // P2549R1 unexpected<E>::error()
 // P2602R2 Poison Pills Are Too Toxic
+// P2609R3 Relaxing Ranges Just A Smidge
```
Comment (CaseyCarter) [inline stl/inc/yvals_core.h]: These should be in the `_HAS_CXX20` section since they're both "always on".

**Review by CaseyCarter — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]
Looks good!

File: stl/inc/xutility [hunk 4]
```diff
@@ -629,69 +638,56 @@ concept contiguous_iterator = random_access_iterator<_It>
...
-    && copy_constructible<_Fn>
-    && invocable<_Fn&, iter_value_t<_It>&>
-    && invocable<_Fn&, iter_reference_t<_It>>
-    && invocable<_Fn&, iter_common_reference_t<_It>>
-    && common_reference_with<
-        invoke_result_t<_Fn&, iter_value_t<_It>&>,
+concept indirectly_unary_invocable =
+    indirectly_readable<_It> && copy_constructible<_Fn> && invocable<_Fn&, _Indirect_value_t<_It>>
+    && invocable<_Fn&, iter_reference_t<_It>> && invocable<_Fn&, iter_common_reference_t<_It>>
+    && common_reference_with<invoke_result_t<_Fn&, _Indirect_value_t<_It>>,
```
Comment (StephanTLavavej) [inline stl/inc/xutility:647]: No change requested: I am uncomfortable with the drive-by reformatting here. The issue is that the reformatting is mixed with the semantic changes, which makes it harder to see what's going on. I would have preferred for this to be done in a separate PR. That said, it's not *too* complicated and we've already reviewed the changes, so let's leave it as-is.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Pull request merged**